attitude-estimator/attitude_estimator.cpp

```diff
@@ -0,0 +1,945 @@
+// Attitude estimator based on the fusion of 3-axis accelerometer, gyroscope and magnetometer data
+// File: attitude_estimator.cpp
+// Author: Philipp Allgeuer <pallgeuer@ais.uni-bonn.de>
+
+// Includes
+#include "attitude_estimator.h"
+#include <cmath>
+
+// Defines
+#define M_2PI 6.2831853071795864769
+
+// Namespaces
+using namespace std;
+using namespace stateestimation;
+
+// Constants
+const double AttitudeEstimator::ACC_TOL_SQ = 1e-12*1e-12;
+const double AttitudeEstimator::QY_NORM_TOL_SQ = 1e-12*1e-12;
+const double AttitudeEstimator::QHAT_NORM_TOL_SQ = 1e-12*1e-12;
+const double AttitudeEstimator::XGYG_NORM_TOL_SQ = 1e-12*1e-12;
+const double AttitudeEstimator::WEZE_NORM_TOL_SQ = 1e-12*1e-12;
+const double AttitudeEstimator::ZGHAT_ABS_TOL = 1e-12;
+
+// Constructor
+AttitudeEstimator::AttitudeEstimator(bool quickLearn)
+{
+	// Initialise the entire class
+	resetAll(quickLearn);
+}
+
+// Reset functions
+void AttitudeEstimator::reset(bool quickLearn, bool resetGyroBias) // Note: Resets the entire class, except for the magnetometer calibration, the acc-only resolution method, and the configuration variables (i.e. the PI gains and the quick learn time)
+{
+	// Reset the estimator state
+	resetState(resetGyroBias);
+
+	// Reset the lambda value to reinitiate quick learning (if requested)
+	if(quickLearn) resetLambda();
+	else setLambda();
+}
+void AttitudeEstimator::resetAll(bool quickLearn) // Note: Resets absolutely everything about the class
+{
+	// Initialise the acc-only resolution method
+	setAccMethod(ME_DEFAULT);
+
+	// Initialise the configuration variables
+	setPIGains(2.20, 2.65, 10.0, 1.25);
+	setQLTime(3.0);
+
+	// Initialise the magnetometer calibration
+	setMagCalib(1.0, 0.0, 0.0);
+
+	// Reset the attitude estimator
+	reset(quickLearn);
+}
+void AttitudeEstimator::resetState(bool resetGyroBias) // Note: Currently used by reset() only
+{
+	// Declare variables
+	int i;
+	
+	// Initialise the attitude estimate
+	setAttitude(1.0, 0.0, 0.0, 0.0); // Resets m_Qhat, m_eulerValid and m_fusedValid internally
+
+	// Update the alternative attitude estimate representations
+	updateEuler(); // Resets m_Ehat and m_eulerValid internally
+	updateFused(); // Resets m_Fhat, m_FhatHemi and m_fusedValid internally
+
+	// Initialise the gyro bias estimate
+	if(resetGyroBias)
+		setGyroBias(0.0, 0.0, 0.0); // Resets m_bhat internally
+
+	// Initialise the remaining size 3 internal variables
+	for(i = 0; i < 3; i++)
```

Comment:
lmao that;s some clever shit

Reply:
yeah, but I'm too dumb to fully understand the code and the math behind it. Btw the original code is from there: https://github.com/AIS-Bonn/attitude_estimator